src/psyclone/domain/lfric/lfric_invoke.py

```diff
@@ -0,0 +1,331 @@
+                if rule.alg_name not in self._alg_unique_qr_args:
+                    self._alg_unique_qr_args.append(rule.alg_name)
+        self._alg_unique_args.extend(self._alg_unique_qr_args)
+        # we also need to work out the names to use for the qr
+        # arguments within the psy layer. These are stored in the
+        # _psy_unique_qr_vars list
+        self._psy_unique_qr_vars = []
+        for call in self.schedule.kernels():
+            for rule in call.qr_rules.values():
+                if rule.psy_name not in self._psy_unique_qr_vars:
+                    self._psy_unique_qr_vars.append(rule.psy_name)
+
+        # lastly, add in halo exchange calls and global sums if
+        # required. We only need to add halo exchange calls for fields
+        # since operators are assembled in place and scalars don't
+        # have halos. We only need to add global sum calls for scalars
+        # which have a gh_sum access.
+        if Config.get().distributed_memory:
+            # halo exchange calls
+            const = LFRicConstants()
+            for loop in self.schedule.loops():
+                loop.create_halo_exchanges()
+            # global sum calls
+            for loop in self.schedule.loops():
+                for scalar in loop.args_filter(
+                        arg_types=const.VALID_SCALAR_NAMES,
+                        arg_accesses=AccessType.get_valid_reduction_modes(),
+                        unique=True):
+                    global_sum = DynGlobalSum(scalar, parent=loop.parent)
+                    loop.parent.children.insert(loop.position+1, global_sum)
+
+    def arg_for_funcspace(self, fspace):
+        ''' Returns an argument object which is on the requested
+        function space. Searches through all Kernel calls in this
+        invoke. Currently the first argument object that is found is
+        used. Throws an exception if no argument exists. '''
+        for kern_call in self.schedule.kernels():
+            try:
+                return kern_call.arguments.get_arg_on_space(fspace)
+            except FieldNotFoundError:
+                pass
+        raise GenerationError(
+            f"No argument found on '{fspace.mangled_name}' space")
+
+    def unique_fss(self):
+        ''' Returns the unique function space *objects* over all kernel
+        calls in this invoke. '''
+        unique_fs = []
+        unique_fs_names = []
+        for kern_call in self.schedule.kernels():
+            kern_fss = kern_call.arguments.unique_fss
+            for fspace in kern_fss:
+                if fspace.mangled_name not in unique_fs_names:
+                    unique_fs.append(fspace)
+                    unique_fs_names.append(fspace.mangled_name)
+        return unique_fs
+
+    @property
+    def operates_on_dofs_only(self):
+        '''
+        :returns: whether or not this Invoke consists only of kernels that \
+                  operate on DoFs.
+        :rtype: bool
```

Comment:
I would suggest adding a blank line below this line and above docstring quotations.

Reply:
346738f